### PR TITLE
relative attention

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- add relative attention
- part of https://github.com/RasaHQ/research/issues/62

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/utils/tensorflow/layers.py [hunk 1]
```diff
@@ -143,263 +142,10 @@ def call(self, x: tf.Tensor) -> tf.Tensor:
...
-
-        # if normalization is done in encoding layers, then it should also be done
-        # on the output, since the output can grow very large, being the sum of
-        # a whole stack of unnormalized layer outputs.
-        return self._layernorm(x)  # (batch_size, seq_len, d_model)
-
-
 class InputMask(tf.keras.layers.Layer):
     def build(self, input_shape: tf.TensorShape) -> None:
-        initializer = tf.keras.initializers.GlorotUniform()
```
Comment (tabergma) [inline rasa/utils/tensorflow/layers.py]: Why did you remove the initializer?

File: rasa/utils/tensorflow/transformer.py [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+import numpy as np
+from rasa.utils.tensorflow.layers import DenseWithSparseWeights
+
+
+# from https://www.tensorflow.org/tutorials/text/transformer
+# and https://github.com/tensorflow/tensor2tensor
+class MultiHeadAttention(tf.keras.layers.Layer):
+    def __init__(
+        self,
+        d_model: int,
```
Comment (tabergma) [inline rasa/utils/tensorflow/transformer.py]: Is it common to name the transformer size `d_model`? If not, can we rename it to `model_size` or `model_dimension`?

File: rasa/utils/tensorflow/transformer.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+        self.attention_dropout_rate = attention_dropout_rate
+        self.unidirectional = unidirectional
+        self.use_key_relative_position = use_key_relative_position
+        self.use_value_relative_position = use_value_relative_position
+        self.relative_length = max_relative_position
+        if self.relative_length is not None:
+            self.relative_length += 1  # include current time
+        self.heads_share_relative_embedding = heads_share_relative_embedding
+
+        assert d_model % self.num_heads == 0
```
Comment (tabergma) [inline rasa/utils/tensorflow/transformer.py]: Should we raise a proper Error if this is not true and add some explanation?

File: rasa/utils/tensorflow/transformer.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        self._wv = DenseWithSparseWeights(units=d_model, use_bias=False)
+
+        self._dense = DenseWithSparseWeights(units=d_model)
+
+        self._create_relative_embeddings()
+
+    def _create_relative_embeddings(self) -> None:
+        """Create relative embeddings."""
+
+        if self.use_key_relative_position or self.use_value_relative_position:
```
Comment (tabergma) [inline rasa/utils/tensorflow/transformer.py]: Minor:
We could get rid of the `else` clauses by first doing
```
relative_embedding_shape = None
self.key_relative_embeddings = None
self.value_relative_embeddings = None
```
and then doing the if part.

File: rasa/utils/tensorflow/transformer.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+
+        x.shape =
+        (batch, num_heads, length, relative_length, depth)
+        or (batch, num_heads, length, relative_length)
+        "Slides" relative embeddings by 45 degree """
+
+        x_dim = len(x.shape)
+
+        if x_dim < 4 or x_dim > 5:
+            raise ValueError("Relative tensor has a wrong shape.")
```
Comment (tabergma) [inline rasa/utils/tensorflow/transformer.py]: Should we mention (1) the shape and (2) state that it should be either of dimensionality 4 or 5?

File: rasa/utils/tensorflow/transformer.py [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+
+        output = self._dense(attention)  # (batch_size, seq_len_q, d_model)
+
+        return output, attention_weights
+
+
+class TransformerEncoderLayer(tf.keras.layers.Layer):
+    def __init__(
+        self,
+        d_model: int,
```
Comment (tabergma) [inline rasa/utils/tensorflow/transformer.py]: same as above

File: rasa/utils/tensorflow/transformer.py [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+        x += ffn_out
+
+        return x  # (batch_size, seq_len, d_model)
+
+
+class TransformerEncoder(tf.keras.layers.Layer):
+    def __init__(
+        self,
+        num_layers: int,
+        d_model: int,
```
Comment (tabergma) [inline rasa/utils/tensorflow/transformer.py]: same as above

File: rasa/utils/tensorflow/transformer.py (hunk 2 above)
Comment (Ghostvv) [reply rasa/utils/tensorflow/transformer.py]: it's just copy pasted from tf tutorial, I will rename

File: rasa/utils/tensorflow/transformer.py (hunk 5 above)
Comment (Ghostvv) [reply rasa/utils/tensorflow/transformer.py]: it's more of an internal method, I added error not make mistake myself

File: rasa/utils/tensorflow/layers.py (hunk 1 above)
Comment (Ghostvv) [reply rasa/utils/tensorflow/layers.py]: because `GlorotUniform` is a standard one

Comment (Ghostvv) [reply rasa/utils/tensorflow/layers.py]: I thought it is cleaner to use standard parameters

File: rasa/utils/tensorflow/transformer.py (hunk 2 above)
Comment (Ghostvv) [reply rasa/utils/tensorflow/transformer.py]: in keras dense layers, it is called `units`
